File: cloud/oauth-server/src/main/java/pers/acp/admin/oauth/token/store/SecurityTokenStoreMemory.java
package pers.acp.admin.oauth.token.store;

import org.springframework.security.oauth2.provider.token.store.InMemoryTokenStore;
import pers.acp.admin.common.constant.CommonConstant;
import pers.acp.admin.oauth.token.LoginLog;
import pers.acp.admin.oauth.token.SecurityTokenStore;
import pers.acp.admin.oauth.vo.LoginLogVO;
import pers.acp.core.CommonTools;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * @author zhang by 08/03/2019
 * @since JDK 11
 */
public class SecurityTokenStoreMemory extends InMemoryTokenStore implements SecurityTokenStore {

    private static ConcurrentLinkedQueue<LoginLog> loginLogConcurrentLinkedQueue = new ConcurrentLinkedQueue<>();

    @Override
    public void storeLoginNum(String appId, String userId) {
        LoginLog loginLog = new LoginLog();
        loginLog.setAppid(appId);
        loginLog.setDate(CommonTools.getDateTimeString(null, CommonConstant.DATE_FORMAT));
        loginLog.setUserid(userId);
        loginLogConcurrentLinkedQueue.add(loginLog);
    }

    @Override
    public List<LoginLogVO> getLoginNum(String appId) throws Exception {
        List<LoginLogVO> loginLogVOList = new ArrayList<>();
        loginLogConcurrentLinkedQueue.forEach(loginLog -> SecurityTokenStoreUtil.parseLoginLogVO(loginLogVOList, loginLog));
        return loginLogVOList;
    }

}

File: cloud/oauth-server/src/main/kotlin/pers/acp/admin/oauth/token/store/SecurityTokenStoreMemory.kt
package pers.acp.admin.oauth.token.store

import org.springframework.security.oauth2.provider.token.store.InMemoryTokenStore
import pers.acp.admin.oauth.token.LoginLog
import pers.acp.admin.oauth.token.SecurityTokenStore
import pers.acp.admin.oauth.vo.LoginLogVo
import pers.acp.core.CommonTools
import pers.acp.core.task.timer.Calculation

import java.util.concurrent.ConcurrentLinkedQueue

/**
 * @author zhang by 08/03/2019
 * @since JDK 11
 */
class SecurityTokenStoreMemory : InMemoryTokenStore(), SecurityTokenStore {

    @Throws(Exception::class)
    override fun storeLoginNum(appId: String, userId: String) {
        loginLogConcurrentLinkedQueue.add(LoginLog(
                appId = appId,
                date = CommonTools.getDateTimeString(null, Calculation.DATE_FORMAT),
                userId = userId
        ))
    }

    @Throws(Exception::class)
    override fun getLoginNum(appId: String): MutableList<LoginLogVo> {
        val loginLogVoList: MutableList<LoginLogVo> = mutableListOf()
        loginLogConcurrentLinkedQueue.forEach { loginLog -> SecurityTokenStoreUtil.parseLoginLogVO(loginLogVoList, loginLog) }
        return loginLogVoList
    }

    companion object {
        private val loginLogConcurrentLinkedQueue = ConcurrentLinkedQueue<LoginLog>()
    }

}
